src/hotspot/cpu/s390/upcallLinker_s390.cpp

```diff
@@ -39,9 +39,10 @@ static int compute_reg_save_area_size(const ABIDescriptor& abi) {
   int size = 0;
   for (int i = 0; i < Register::number_of_registers; i++) {
     Register reg = as_Register(i);
-    // Z_SP and Z_R14 saved/restored by prologue/epilogue
-    if (reg == Z_SP || reg == Z_R14) continue;
-    if (!abi.is_volatile_reg(reg)) {
+    // Z_SP saved/restored by prologue/epilogue
+    if (reg == Z_SP) continue;
+    // though Z_R6 is argument register it is a saved register
+    if (!abi.is_volatile_reg(reg) || reg == Z_R6) {
```

Comment:
So, is the prior assumption that all argument registers are also volatile incorrect?

If so, I think it would be better to change `ABIDescriptor::is_volatile_reg` ([1]) to only look at the `_integer_additional_volatile_registers` list (maybe rename it too), and then list all the volatile regs in  LinuxS390CallArranger explicitly, excluding R6 ([2]). That way all the information about which regs are volatile is in one place (LinuxS390CallArranger).

[1]: https://github.com/openjdk/jdk/pull/14801/files#diff-7096e1975de20baa3219d616506f26ba2b4500bf5ad28c331e3d6049a32a461eR39-R42
[2]: https://github.com/openjdk/jdk/pull/14801/files#diff-09da9016992b04bab73b6cc6aad8ca719e86c90c398af4e000583c1f8220a99bR73

Reply:
Yes, s390x ABI says R6 is an argument register as we as non-volatile register.
It make sense, I'll make suggested changes.